src/Sarif/Visitors/ReformattingVisitor.cs

```diff
@@ -43,23 +45,129 @@ public override Run VisitRun(Run node)
+        public override PhysicalLocation VisitPhysicalLocation(PhysicalLocation node)
         {
-            Uri uri;
+            // If we have a binary region, there is no work to do
+            if (node.Region == null ||
+                (node.Region.ByteOffset > 0 || node.Region.ByteLength > 0))
+            {
+                goto Exit;
+            }
+
+            // If we already have a snippet, let's assume it is correct
+            if (node.Region.Snippet != null)
+            {
+                goto Exit;
+            }
 
-            if (!Uri.TryCreate(key, UriKind.RelativeOrAbsolute, out uri))
+            // If the user hasn't specified code snippet insertion, there is no work to do
+            bool workToDo = false;
+            workToDo |= node.Region != null && _dataToInsert.Includes(OptionallyEmittedData.CodeSnippets);
+
+            if (!workToDo)
+            {
+                goto Exit;
+            }
+
+            // If we have an absolute URL specified as a file that 
+            // exist on disk, there is work to do
+            if (node.FileLocation.Uri.IsAbsoluteUri &&
+                node.FileLocation.Uri.LocalPath != null &&
+                File.Exists(node.FileLocation.Uri.LocalPath))
+            {
+                Region region = node.Region;
+                string fileText = GetFileText(node.FileLocation.Uri);
+                GetCompleteFirstLineAssociatedWithRegion(region, fileText);
+            }
+
+            Exit:
+            return base.VisitPhysicalLocation(node);
+        }
+
+        internal static void GetCompleteFirstLineAssociatedWithRegion(Region region, string fileText)
+        {
+            if (region.IsBinaryRegion)
+            {
+                return;
+            }
+
+            var lineIndex = new NewLineIndex(fileText);
+
+            region.Snippet = new FileContent();
+
+            int startLine;
+            LineInfo lineInfo;
+
+            if (region.StartLine == 0)
+            {
+                lineInfo = lineIndex.GetLineInfoForOffset(region.CharOffset);
+                startLine = lineInfo.LineNumber;
+            }
+            else
+            {
+                lineInfo = lineIndex.GetLineInfoForLine(region.StartLine);
+                startLine = region.StartLine;
+            }
+
+            int lineOffset = lineInfo.StartOffset;
+            int lineLength;
+
+            if (startLine == lineIndex.MaximumLineNumber)
+            {
+                // End of file. Compute length using file length
+                lineLength = fileText.Length - lineOffset;
+            }
+            else
+            {
+                // Not at end of file. Get the offset from the
+                // line after this one, then compute length from there.
+                lineInfo = lineIndex.GetLineInfoForLine(startLine + 1);
+                lineLength = lineInfo.StartOffset - lineOffset;
```

Comment:
Won't this calculation include the newline sequence?

Reply:
yes. is there a concern? that's why we have to trim it later. 

---
In reply to: [202106386](https://github.com/Microsoft/sarif-sdk/pull/942#discussion_r202106386) [](ancestors = 202106386)

Reply:
Not a concern if it gets trimmed. I didn't understand well enough how the code fit together to realize that it would be trimmed later. A comment to that effect wouldn't come amiss.

Reply:
I will make this change. I would appreciate your approval soon as this change is getting large and blocking other work.

---
In reply to: [202437257](https://github.com/Microsoft/sarif-sdk/pull/942#discussion_r202437257) [](ancestors = 202437257)